### PR TITLE
Model Fitter Refactor - Add tests to Covariance Service and main server

### PR DESCRIPTION
**Pull request by AXSaketh**
Contributes to issues/ax-mcp-internal/40

This took 3-ish days to get setup and all.

---

> [!NOTE]
> **Medium Risk**
> Stricter covariance validation may cause previously-accepted API responses to be treated as failures, and the switch to a long-lived client could expose lifecycle/connection issues; the rest of the PR is test-only.
> 
> **Overview**
> Tightens `CovarianceService` by reusing a singleton `AxiomaticAPIClient` instance (instead of creating a client per call) and by making `_has_valid_covariance` reject covariance outputs that are non-square, non-symmetric, or not positive-(semi)definite.
> 
> Adds a new test suite covering the covariance API request/response shape, the `compute_parameter_covariance` MCP tool wiring (mocked service call), `CovarianceService` validation behavior, and extensive `data_file_utils` file/mapping/error handling cases via new `tests/conftest.py` fixtures.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 30841a3be5e1405f5a149fe5211f7cf84d07683f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: axiomatic_mcp/servers/axmodelfitter/services/covariance_service.py [hunk 1]
```diff
@@ -85,7 +100,17 @@ def _has_valid_covariance(self, response: dict) -> bool:
         """
         robust_cov = response.get("sandwich_covariance")
         classical_cov = response.get("inverse_hessian_covariance")
-        return (isinstance(robust_cov, list) and len(robust_cov) > 0) or (isinstance(classical_cov, list) and len(classical_cov) > 0)
+        is_square = lambda mat: isinstance(mat, list) and all(isinstance(row, list) and len(row) == len(mat) for row in mat)
+        is_symmetric = lambda mat: is_square(mat) and all(np.isclose(mat[i][j], mat[j][i]) for i in range(len(mat)) for j in range(len(mat)))
+        is_positive_definite = lambda mat: is_square(mat) and np.all(np.linalg.eigvals(np.array(mat)) >= 0)
```
Comment (carlosarriAX) [inline axiomatic_mcp/servers/axmodelfitter/services/covariance_service.py]: for symetric matrices use `np.linalg.eigvalsh` instead of `np.linalg.eigvals`. also is this supposed to be positive definite > 0? or semi-positive >=0

File: tests/servers/axmodelfitter/api/compute_parameter_covariance_tests.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+"""Tests for the ax-api /digital-twin/compute_parameter_covariance endpoint."""
+
+import json
```
Comment (carlosarriAX) [inline tests/servers/axmodelfitter/api/compute_parameter_covariance_tests.py]: json and requests are not used

File: tests/servers/axmodelfitter/api/compute_parameter_covariance_tests.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+@pytest.fixture
+def endpoint():
+    return "/digital-twin/compute-parameter-covariance"
+
+@pytest.fixture
+def client():
+    return AxiomaticAPIClient()
+
+# TODO Why does the request body need to contain lowercase false for json? Didn't feel like tracing the object back all the way. 
+def test_compute_parameter_covariance_valid_request(client, endpoint, simple_request_body_parameter_covariance):
```
Comment (carlosarriAX) [inline tests/servers/axmodelfitter/api/compute_parameter_covariance_tests.py]: this test hits the real API. could you add `@pytest.mark.integration`, so they are identified as integration tests.

File: tests/servers/axmodelfitter/services/covariance_service_tests.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    """Return a covariance matrix that is not positive definite."""
+    return [[-0.1, 0.01], [0.01, -0.2]]
+
+@pytest.mark.asyncio
+async def test_covariance_service(covariance_service, nonlinear_request_body_parameter_covariance):
+    """Ensure a valid request is successful."""
+    result = await covariance_service.compute_covariance(nonlinear_request_body_parameter_covariance)
+    assert result["success"] == True
+    
+def test_has_valid_covariance(covariance_service, valid_covariance_matrix, invalid_covariance_matrix_non_square, invalid_covariance_matrix_non_symmetric, invalid_covariance_matrix_non_positive_definite):
```
Comment (carlosarriAX) [inline tests/servers/axmodelfitter/services/covariance_service_tests.py:33]: It would be cleaner to split this into separate test. Also consider adding a test when both matrices are none.

File: axiomatic_mcp/servers/axmodelfitter/templates/analytical_trigonometric.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+SOURCE = """def y(t, amplitude, frequency, phase, offset):
+return amplitude * jnp.sin(2 * jnp.pi * frequency * t + phase) + offset"""
```
Comment (AdriantrejoAx) [inline axiomatic_mcp/servers/axmodelfitter/templates/analytical_trigonometric.py]: I think you need to indent this

File: tests/servers/axmodelfitter/api/compute_parameter_covariance_tests.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+"""Tests for the ax-api /digital-twin/compute_parameter_covariance endpoint."""
```
Comment (AdriantrejoAx) [inline tests/servers/axmodelfitter/api/compute_parameter_covariance_tests.py:1]: Why do you need api tests? These should be inside ax-stack in the integration tests

File: tests/servers/axmodelfitter/data_file_utils_tests.py [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+    load_data_file,
+    resolve_data_input,
+    resolve_output_data_only,
+    transform_file_to_optimization_format,
+    validate_column_mapping,
+    validate_file_access,
+)
+
+
+# ---------------------------------------------------------------------------
```
Comment (AdriantrejoAx) [inline tests/servers/axmodelfitter/data_file_utils_tests.py]: Please separate the fixtures in other directory and remove the ----- AI comments

File: axiomatic_mcp/servers/axmodelfitter/services/covariance_service.py [hunk 8]
```diff
@@ -17,6 +18,21 @@ class CovarianceService(SingletonBase):
     classical inverse Hessian approach.
     """
 
+    def __init__(self):
+        if getattr(self, "_client", None) is not None:
+            return
+        self._client = AxiomaticAPIClient()
+        atexit.register(self.close)
```
Comment (carlosarriAX) [inline axiomatic_mcp/servers/axmodelfitter/services/covariance_service.py]: There is only one client. I would add a flag to track whether atexit has been registered, so it doesnt register multiple unecessary times.

File: tests/servers/axmodelfitter/services/covariance_service_tests.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
+"""Tests for axmodelfitter/services/covariance_service.py"""
+
+import pytest
+import pytest_asyncio
+
+from axiomatic_mcp.servers.axmodelfitter.services.covariance_service import CovarianceService
+
+@pytest.fixture
```
Comment (AdriantrejoAx) [inline tests/servers/axmodelfitter/services/covariance_service_tests.py:8]: You are making a real HTTP call to the API endpoint, you need to mock the client

**Review by AdriantrejoAx — CHANGES_REQUESTED** [review]

File: axiomatic_mcp/servers/axmodelfitter/services/covariance_service.py (hunk 8 above)
Comment (AXSaketh) [reply axiomatic_mcp/servers/axmodelfitter/services/covariance_service.py]: I see, if CovarianceService is created multiple times after being closed in between, atexit.register will happen again.

**Comment by AdriantrejoAx** [comment]
Closing due to inactivity
